Deploy nse.gene-expression-volume [BBPP82-482]

lguyot opened the pull request:
I have successfully tested this PR on BB5 (but not the deployment yet).

lguyot committed:
Deploy nse.gene-expression-volume [BBPP82-482]

matz-e commented on var/spack/repos/builtin/packages/gene-expression-volume/package.py:
```diff
@@ -0,0 +1,34 @@
+    depends_on('py-pyaml@5.3.1:', type=('build', 'run'))
+    depends_on('py-pytest', type='test')
+    depends_on('py-mock', type='test')
+
```
```suggestion

    def patch(self):
        filter_file('"tensorflow==[0-9.]+",', '', 'setup.py')
```

lguyot replied:
@matz-e Thanks for looking into this. I have tried your suggestion, but a "new" problem arises then:

```
 File "/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/libraries/2021-01-06/linux-rhel7-x86_64/gcc-9.3.0/py-setuptools-50.3.2-afe2g4/lib/python3.8/site-packages/pkg_resources/__init__.py", line 770, in resolve
    raise DistributionNotFound(req, requirers)
pkg_resources.DistributionNotFound: The 'lpips_tf@ git+http://github.com/alexlee-gk/lpips-tensorflow.git#egg=lpips_tf' distribution was not found and is required by gene-expression-volume
``` 

I thought that the following could help:

```
def patch(self):
    filter_file('"lpips_tf.+=lpips_tf",', '', 'setup.py')
    filter_file('"tensorflow==[0-9.]+",', '', 'setup.py')
```
but then I get

```
  File "/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/libraries/2021-01-06/linux-rhel7-x86_64/gcc-9.3.0/py-setuptools-50.3.2-afe2g4/lib/python3.8/site-packages/pkg_resources/__init__.py", line 770, in resolve
    raise DistributionNotFound(req, requirers)
pkg_resources.DistributionNotFound: The 'tensorflow>=2.4.0' distribution was not found and is required by atlalign
```

which bewildered me.

GianlucaFicarelli commented on var/spack/repos/builtin/packages/py-types-pyaml/package.py:
```diff
@@ -0,0 +1,18 @@
+from spack import *
+
+
+class PyTypesPyaml(PythonPackage):
```
I know it's still a draft, but note that the name of the package and the class should be `py-types-pyyaml` and `PyTypesPyyaml` instead of `py-types-pyaml` and `PyTypesPyaml`.

lguyot replied:
Thanks Gianluca for spotting this. I'll fix it in the next commit.

lguyot replied:
Done in the next commit.

lguyot committed:
Patches gene-expression-volume, py-atlalign and its dependent py-imgaug - Addresses naming issues raised by Gianluca

The main issue is that some packages <packages> referred to in the setup.py file
of <parent-package> are not visible for <parent-package> whereas the <packages>
python modules are actually installed.

The patch consists in removing the <packages> dependencies
from <parent-package>'s setup.py file (prunes install_requires).

The occurences of this problem are:
- parent-package = atlalign, packages = [tensorflow]
- parent-package = imgaug, packages = [opencv-python]
- parent-package = gene-expression-volume, packages = [tensorflow, lpips_tf]

lguyot committed:
Fixes flake8 errors (trailing spaces and blank lines)

lguyot committed:
Fixes flawed reference to py-pyyaml in gene-expression-volume/package.py

matz-e commented on var/spack/repos/builtin/packages/py-atlalign/package.py:
```diff
@@ -32,3 +32,6 @@ class PyAtlalign(PythonPackage):
     patch('lpips.patch', when='@0.6.0')
+
+    def patch(self):
+        filter_file('"tensorflow>=[0-9.]+",', '', 'setup.py')
```
I'm not sure if this is enough to trigger a rebuild of the package.

lguyot replied:
Thanks for raising this point. So I'll use the same trick as the one of `py-tensorflow-addons/package.py`:  add a new version with the same `sha256`. See next commit.

lguyot committed:
Deploy gene-expression-volume and bumps versions of py-atlalign and py-imgaug to trigger their re-builds

lguyot committed:
Fixes flake8 errors and wrong reference to py-type-pyyaml

matz-e commented:
Retest this please

lguyot commented:
Retest this please

GianlucaFicarelli commented on var/spack/repos/builtin/packages/py-atlalign/package.py:
```diff
@@ -12,6 +12,8 @@ class PyAtlalign(PythonPackage):
     url = "https://files.pythonhosted.org/packages/74/b8/4382a758e4ccf5de26e35fb2b33863aea27b99e31e1ce9b4bc1e6bb593ef/atlalign-0.6.0.tar.gz"
 
+    # The first version below just serves to trigger a rebuild!
+    version('0.6.0.20210902', sha256='ecd74c89ecdca0a115252fa662e41b272dff2cdc3ea0d459daa5b7a672070b7c')
```
With this version `0.6.0.20210902`, I think that spack fails to fetch the package because it would try to download https://files.pythonhosted.org/packages/74/b8/4382a758e4ccf5de26e35fb2b33863aea27b99e31e1ce9b4bc1e6bb593ef/atlalign-0.6.0.20210902.tar.gz
If it's temporary, you can try with
```suggestion
    version(
        '0.6.0.20210902',
        sha256='ecd74c89ecdca0a115252fa662e41b272dff2cdc3ea0d459daa5b7a672070b7c',
        url='https://files.pythonhosted.org/packages/74/b8/4382a758e4ccf5de26e35fb2b33863aea27b99e31e1ce9b4bc1e6bb593ef/atlalign-0.6.0.tar.gz'
        )
```
Or just remove the build directory? (but it has to be done manually)

lguyot replied:
@GianlucaFicarelli Many thanks for stepping in. You made me realize that, indeed, I messed up with urls. For the re-build hack to work as in `py-tensorflow-addons/package.py`, it suffices to add `url=url,  expand=False` to the parameters of `version`. I completely forgot that. What you suggest is, of course, also valid, but more verbose.

Many thanks again for  your hint. I have made changes in the next commit accordingly.

lguyot committed:
Fixes urls of the packages to be re-built (py-atlalign and py-imgaug)

matz-e commented on var/spack/repos/builtin/packages/py-imgaug/package.py:
```diff
@@ -12,6 +12,8 @@ class PyImgaug(PythonPackage):
     url = "https://files.pythonhosted.org/packages/4f/96/c55f2985b464ea377238e61026dfb1da54d8edc1534c649e70077e94c642/imgaug-0.2.9.tar.gz"
 
+    # The first version below just serves to trigger a rebuild!
+    version('0.2.9.20210902', url=url, sha256='42b0c4c8cbe197d4f5dbd33960a1140f8a0d9c22c0a8851306ecbbc032092de8', expand=False)
```
Isn't this a proper `PythonPackage`? In that case we need to grab the sources, not a full wheel to be installed.
```suggestion
    version('0.2.9.20210902', sha256='42b0c4c8cbe197d4f5dbd33960a1140f8a0d9c22c0a8851306ecbbc032092de8',)
```

lguyot replied:
Yes, you are right. I was slowly coming to this. Fixed in the next commit.

lguyot committed:
Removes the --expand=False options from py-atlalign and py-imgaug package.py files: they cause install problems

GianlucaFicarelli commented on var/spack/repos/builtin/packages/gene-expression-volume/package.py:
```diff
@@ -0,0 +1,38 @@
+    depends_on('py-scikit-image@0.017.2:', type=('build', 'run'))
+    depends_on('py-types-pyyaml@5.4.0:', type=('build', 'run'))
+    depends_on('py-voxcell@2.7.4', type=('run', 'build'))
+    depends_on('py-pyaml@5.3.1:', type=('build', 'run'))
```
This should be `py-pyyaml` (`py-pyaml` exists but it's a different package).
```suggestion
    depends_on('py-pyyaml@5.3.1:', type=('build', 'run'))
```

lguyot replied:
Indeed, that's awkward! I'll address this problem in a new PR, together with all the new issues you have identified.

GianlucaFicarelli commented on var/spack/repos/builtin/packages/gene-expression-volume/package.py:
```diff
@@ -0,0 +1,38 @@
+    depends_on('py-click@7.0:', type=('build', 'run'))
+    depends_on('py-nptyping@1.0.1:', type=('build', 'run'))
+    depends_on('py-numpy@1.15.0:', type=('build', 'run'))
+    depends_on('py-scikit-image@0.017.2:', type=('build', 'run'))
```
Minor and probably working in either way:
```suggestion
    depends_on('py-scikit-image@0.17.2:', type=('build', 'run'))
```

lguyot replied:
To be fixed in the next PR.

GianlucaFicarelli commented on var/spack/repos/builtin/packages/gene-expression-volume/package.py:
```diff
@@ -0,0 +1,38 @@
+    depends_on('py-numpy@1.15.0:', type=('build', 'run'))
+    depends_on('py-scikit-image@0.017.2:', type=('build', 'run'))
+    depends_on('py-types-pyyaml@5.4.0:', type=('build', 'run'))
+    depends_on('py-voxcell@2.7.4', type=('run', 'build'))
```
Shouldn't be like this?
```suggestion
    depends_on('py-voxcell@2.7.4:2.999', type=('run', 'build'))
```
However the result is the same, because there are only py-voxcell 2.7.4 and 3.0.0.

lguyot replied:
Addressed in the next PR.

GianlucaFicarelli commented on var/spack/repos/builtin/packages/py-types-pyyaml/package.py:
```diff
@@ -0,0 +1,17 @@
+    """Typing stubs for PyYAML."""
+
+    homepage = "https://github.com/python/typeshed"
+    url      = "https://files.pythonhosted.org/packages/3c/96/22a02c905aec42d8f7f7c48b7a00d68019edd4768e3fb94b754966d07870/types-PyYAML-5.4.6.tar.gz"
```
Is there any reason to set the url to the static url with the checksum?
To avoid adding a url specific to a version, and letting spack to resolve the url for different versions, it would be possible to use one of
- https://pypi.io/packages/source/t/types-pyyaml/types-PyYAML-5.4.6.tar.gz
- https://pypi.org/packages/source/t/types-pyyaml/types-PyYAML-5.4.6.tar.gz
- https://files.pythonhosted.org/packages/source/t/types-pyyaml/types-PyYAML-5.4.6.tar.gz

In https://warehouse.readthedocs.io/api-reference/integration-guide.html#querying-pypi-for-package-urls they suggest to use json, but I think it's not possible with spack (but in the upstream release it's possible to use the `pypi` keyword).

lguyot replied:
Addressed in the next PR.
